c/httpserver.c

```diff
@@ -5503,61 +5747,218 @@ static int httpHandleTCP(STCBase *base,
+    muxSocket->userData = muxExtension;
 
-      if (conversation->shouldClose) {
-        /* immediately unregister the socketExtension so we can't enqueue more than one CLOSE_CONVERSATION request */
-        stcReleaseSocketExtension(base, peerExtension);
+    HttpConversation *httpConversation = makeHttpConversation(handlerExtension,(HttpServer*)(module->data));
 
-        serializeConsiderCloseEnqueue(conversation,FALSE);
+    stcRegisterSocketExtension(server->base, handlerExtension, STC_MODULE_JEDHTTP);
+    stcRegisterSocketExtension(server->base, muxExtension, STC_MODULE_JEDHTTP);
 
-        break; /* end peer socket processing */
-      }
+    entry = (PipeSocketEntry*)safeMalloc(sizeof(PipeSocketEntry),"PipeSocketEntry");
+    entry->uniqueID = uniqueID;
+    entry->demuxSocket = demuxSocket;
+    entry->handlerSocket = handlerSocket;
+    entry->muxSocket = muxSocket;
+    zowelog(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_DEBUG2, "PSEntry at 0x%p\n",entry );
+    zowedump(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_DEBUG2, (char*)entry,sizeof(PipeSocketEntry));
+    muxExtension->protocolHandler = entry;
 
-      zowelog(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_DEBUG3, "peerExtension at 0x%x, httpConversation at 0x%x\n", 
-              peerExtension, conversation);
+    htIntPut(server->syntheticPipeSockets,uniqueID,entry);
+  }
+  return entry; 
+}
 
-#if defined(__ZOWE_OS_ZOS) || defined(USE_RS_SSL) || defined(USE_RS_TLS)
-      int sxStatus = sxUpdateTLSInfo(peerExtension,
-                                     1); /* prevent multiple ioctl calls on repeated reads */
-      if (0 != sxStatus) {
-        zowelog(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_DEBUG3, "error from sxUpdateTLSInfo: %d\n", sxStatus);
-      } else if ((RS_TLS_WANT_TLS & peerExtension->tlsFlags) &&
-                 (0 == (RS_TLS_HAVE_TLS & peerExtension->tlsFlags)))
-      {
-        if (0 == (tlsWarnCounter % TLS_WARN_FREQUENCY)) {
-          zowelog(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_WARNING, "Connection is insecure! TLS needed but not found on socket.\n");
-        }
-        tlsWarnCounter++;
-      } else if ((RS_TLS_WANT_PEERCERT & peerExtension->tlsFlags) &&
-                 (0 == (RS_TLS_HAVE_PEERCERT & peerExtension->tlsFlags)))
-      {
-        zowelog(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_DEBUG3, "Connection is insecure! Peer certificate wanted but not found.\n");
-      }
-#endif
+/* this is called when PIPE has data 
+ */
 
-      zowelog(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_DEBUG3, "handle read JEDHTTP convo\n");
-      /* if successful, these methods enqueue work */
-      if (conversation->wsSession){
-        doWSReadWork(conversation,READ_BUFFER_SIZE);
-      } else {
-        doHttpReadWork(conversation,READ_BUFFER_SIZE);
+/* this method *only* runs in the STCBase main thread, putting and effective mutex 
+   on the pipes that talk up/down to process that owns this httpserver */
+static int httpHandlePipe(STCBase *base,
```

Comment:
This is the basis of single user Zowe/ZSS.   It does not have a TCP Header.    At no place can I pick this up, ie, no code is written below the level of normal socket or pipe code.   The data structure in question is below and associates the tunnelled conversation to the "remotClientID" which is as unique as the tcp connections made to the far-side (non-ZSS) side of the tunnel.    I have further work to go, there is no "production" code that uses this.   I am trying to get to single user mode being robust enough for demos, and make more fixes.    

typedef struct TCPFragment{
  uint32_t magic;        /* D0BED0BE */
  uint16_t headerLength;
  uint8_t  flags;
  uint8_t  reserved;
  uint32_t type;
  uint32_t remoteClientID; /* proxy for client socket at far end of tunnel */
  /* offset 10 */
  uint32_t payloadLength;
  char     optionalHeaderData[];
} TCPFragment;

Reply:
But who generates those headers? At this point I don't understand the full lifecycle of this, i.e. who (what kind of application) connects to ZSS in this mode, how the data is formed on the client end.

And I guess I got confused because it's called "TCP fragment", so I thought it was using the TCP protocol in some form.

Reply:
Talked to Joe, my questions have been answered.